src/client/Microsoft.Identity.Client/Internal/IdToken.cs

```diff
@@ -21,42 +26,39 @@ internal static class IdTokenClaim
         public const string Upn = "upn";
     }
 
-    [JsonObject]
-    [Preserve(AllMembers = true)]
+    internal static class JsonClaimValueTypes
+    {
+        public const string Json = "JSON";
+        public const string JsonArray = "JSON_ARRAY";
+        public const string JsonNull = "JSON_NULL";
+    }
+
     internal class IdToken
     {
-        [JsonProperty(PropertyName = IdTokenClaim.Issuer)]
-        public string Issuer { get; set; }
+        private const string DefaultIssuser = "LOCAL AUTHORITY";
 
-        [JsonProperty(PropertyName = IdTokenClaim.ObjectId)]
-        public string ObjectId { get; set; }
+        public string ObjectId { get; private set; }
 
-        [JsonProperty(PropertyName = IdTokenClaim.Subject)]
-        public string Subject { get; set; }
+        public string Subject { get; private set; }
 
-        [JsonProperty(PropertyName = IdTokenClaim.TenantId)]
-        public string TenantId { get; set; }
+        public string TenantId { get; private set; }
 
-        [JsonProperty(PropertyName = IdTokenClaim.Version)]
-        public string Version { get; set; }
+        public string PreferredUsername { get; private set; }
 
-        [JsonProperty(PropertyName = IdTokenClaim.PreferredUsername)]
-        public string PreferredUsername { get; set; }
+        public string Name { get; private set; }
 
-        [JsonProperty(PropertyName = IdTokenClaim.Name)]
-        public string Name { get; set; }
+        public string Upn { get; private set; }
 
-        [JsonProperty(PropertyName = IdTokenClaim.HomeObjectId)]
-        public string HomeObjectId { get; set; }
+        public string GivenName { get; private set; }
 
-        [JsonProperty(PropertyName = IdTokenClaim.Upn)]
-        public string Upn { get; set; }
+        public string FamilyName { get; private set; }
 
-        [JsonProperty(PropertyName = IdTokenClaim.GivenName)]
-        public string GivenName { get; set; }
+        public string GetUniqueId()
+        {
+            return ObjectId ?? Subject;
+        }
 
-        [JsonProperty(PropertyName = IdTokenClaim.FamilyName)]
-        public string FamilyName { get; set; }
+        public ClaimsPrincipal ClaimsPrincipal { get; private set; }
 
         public static IdToken Parse(string idToken)
```

Comment:
>Parse [](http://example.com/codeflow?start=30&length=5)

Very expensive operations. it should be ensured this is only done if really needed. Overall also should be considered to use SPAN's and reduce the amount of strings created.

Reply:
Would be great, but we need to move to .net core or net5 only :) ... 
We aren't doing a LOT of extra work, because we were already parsing the idtoken and extracting `tid`, `sub`, `name`, etc. Now we are parsing ALL the claims. 

The problem is that each tenant profile will have 1 id token so 10 profiles => 10 token parsings. I'll look at how to eliminate that. 

This is still work in progress, so I'll mark the PR as "draft".

Reply:
Implemented the lazy approach, id tokens will only be parsed when accessing things like `TenantProfile.ObjectId` or `TenantProfile.ClaimsPrincipal` etc.